syntax = "proto3";

package tests;

//message sending and receiving service
service Chat {
  //method to "stablish" a trusted connection
  rpc HandShake (Act1Request) returns (basicMsg) {}
  //method to send a msg
  rpc Dispatch(FullMsg) returns (basicMsg) {}
  //method to recover the msgs received
  rpc Retrieve(basicMsg) returns (stream FullMsg) {}
}

//service to get the complete list of clients
service Clients {
  //method to show all the clients
  rpc GetClients (basicMsg) returns (stream Act1Request) {}
}

//service to get all messages sent by the client
service GetAll {
  //method to retrieve every sent message by a user
  rpc GetAllSent (basicMsg) returns (stream FullMsg) {}
}

// The request message containing the user's name.
message Act1Request {
  string Username = 1;
}

//a basic message with just the necesary, almost like a control message
message basicMsg {
    string From = 1;
    string Msg = 2;
}

//the message that will contain the data shared between clients
message FullMsg {
  string From = 1;
  string To = 2;
  string Msg = 3;
  string Date = 4;
}